### PR TITLE
Refactor exportable media protocol

### PR DESCRIPTION
**Pull request by SergioEstevao**
This PR refactor the `MediaExporter` protocol to make all the exporters to have a common interface to invoke them. 

This converts all the exporters to have a common return type on the callback and then allows them to be invoked using the same code and the answer to be analysed with a common code.

This is a preliminary step to allow common interface on all exporter to report progress.

To test:
 - Check in the app the following features:
 - Media Library
 - Gravatar selection
 - Site Icon selection
 - Aztec editor, insert images

### CODE REVIEW

File: WordPress/Classes/Utility/Media/MediaImageExporter.swift [hunk 1]
```diff
@@ -59,6 +55,42 @@ class MediaImageExporter: MediaExporter {
     ///
     fileprivate let defaultImageFilename = "image"
 
+    private let image: UIImage?
+    private let data: Data?
+    private let url: URL?
+    private let filename: String?
+
+    init(image: UIImage, filename: String?) {
```
Comment (frosty) [inline WordPress/Classes/Utility/Media/MediaImageExporter.swift]: I think these could all delegate to another initialiser that takes all the values as parameters – would reduce each of these to one line?

File: WordPress/Classes/Utility/Media/MediaExporter.swift [hunk 2]
```diff
@@ -1,37 +1,32 @@
 import Foundation
 
-/// General MediaExport protocol, and its requirements.
+/// The MediaExport class represents the result of an MediaExporter.
 ///
-protocol MediaExport {
+@objc class MediaExport: NSObject {
```
[the remark this answers is not in the document]
Comment (SergioEstevao) [reply WordPress/Classes/Utility/Media/MediaExporter.swift]: Good catch, this was a left over from a different implementation. Going to remove it.

File: WordPress/Classes/Utility/Media/MediaThumbnailExporter.swift [hunk 3]
```diff
@@ -172,21 +186,20 @@ class MediaThumbnailExporter: MediaExporter {
     /// Export a thumbnail for a known image at the URL, using self.options for ImageExporter options.
     ///
     fileprivate func exportImageThumbnail(at url: URL, onCompletion: @escaping OnThumbnailExport, onError: @escaping OnExportError) {
-        let exporter = MediaImageExporter()
-        exporter.mediaDirectoryType = .cache
+        let exporter = MediaImageExporter(url: url)
+        exporter.mediaDirectoryType = .temporary
```
[the remark this answers is not in the document]
Comment (SergioEstevao) [reply WordPress/Classes/Utility/Media/MediaThumbnailExporter.swift]: This took me a while to understand, so the idea is that the thumbnail are create on temporary folder and then moved to the cache folder. This avoid collisions on names and clears the temporary files automatically.

If you check the callback on success for both methods them calls : ```self.exportImageToThumbnailCache(export, onCompletion: onCompletion, onError: onError)``` this moves the generated thumbnail to the correct folder that is selected for the thumbnail exporter.

File: WordPress/Classes/Utility/Media/MediaImageExporter.swift (hunk 1 above)
Comment (SergioEstevao) [reply WordPress/Classes/Utility/Media/MediaImageExporter.swift]: Good suggestion, going to implement that.

File: WordPress/Classes/Utility/Media/MediaURLExporter.swift [hunk 4]
```diff
@@ -3,25 +3,13 @@ import MobileCoreServices
 
 /// Media export handling of URLs.
 ///
-class MediaURLExporter: MediaExporter {
+class MediaURLExporter: NSObject, MediaExporter {
```
Comment (frosty) [inline WordPress/Classes/Utility/Media/MediaURLExporter.swift]: Can we remove the NSObject here as well, then?

File: WordPress/Classes/Utility/Media/MediaImageExporter.swift (hunk 1 above)
Comment (SergioEstevao) [reply WordPress/Classes/Utility/Media/MediaImageExporter.swift]: Actually this is not a good idea, for example if you set an UIImage, you should set a Data, the same applies to the URL.

Comment (frosty) [reply WordPress/Classes/Utility/Media/MediaImageExporter.swift]: I'm not sure I follow? I was thinking something like this:

```
private init(image: UIImage?, filename: String?, data: Data?, url: URL?) {
    self.image = image
    self.filename = filename
    self.data = data
    self.url = url
}

convenience init(image: UIImage, filename: String?) {
    self.init(image: image, filename: filename, data: nil, url: nil)
}

convenience init(url: URL) {
    self.init(image: nil, filename: nil, data: nil, url: url)
}

convenience init(data: Data, filename: String?) {
    self.init(image: nil, filename: filename, data: data, url: nil)
}
```

**Comment by SergioEstevao** [comment]
@frosty ready for another look.

**Pull request merged**